### PR TITLE
Pseudocode for First Non Repeating character in Stream

### PR DESCRIPTION
**Pull request by stellarbeam**

### CODE REVIEW

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    \begin{algorithm}
+    
+      \caption{Find the first non-repeating character when each character of the stream arrives}
+      
+      \begin{algorithmic}[1]
+        \Ensure {All characters of string $str$ are lowercase alphabets}
+        \Statex
+        \Function{First\_NonRepeating\_Char}{$str$}
+            \State{Let $Q$ be a queue of characters}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: I think it's better to replace all the `$q$` to `$queue$`. This way the user can constantly visualize the queue.

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+\begin{document}
+
+    \begin{algorithm}
+    
+      \caption{Find the first non-repeating character when each character of the stream arrives}
+      
+      \begin{algorithmic}[1]
+        \Ensure {All characters of string $str$ are lowercase alphabets}
+        \Statex
+        \Function{First\_NonRepeating\_Char}{$str$}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: Avoid using camelcase. Instead, use `Non\_Repeating`.

Moreover, the title is in all caps which is not so appealing. Enclose the title in Math mode.

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            \State{Let $Q$ be a queue of characters}
+            \State{Let $count$ be an array of size 26, which keeps track of frequency of alphabets.}
+            \State{Initialize the $count$ array with 0 for each alphabet}
+            \newline
+            \State{Scan $str$ from left to right and,}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: Replace `$ch$` by `$char$`

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            \For{each character $ch$}
+                \State{Push $ch$ to $Q$}
+                \State{Increment frequency count of $ch$ by 1}
+                
+                \While{$Q$ is not empty}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: Use `\textbf{not empty, }` or refer to the macro in the code of Wildcard Pattern Matching (Linked below)

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+                    \If{Frequency of top element of $Q$ is greater than 1}
+                        \State{Pop the top element of $Q$}
+                    \Else 
+                        \State{Break this loop}
+                    \EndIf
+                \EndWhile
+                \If{$Q$ is not empty}
+                    \State{Print the top element of $Q$}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: It's better to be as concise as possible. use `\textbf{Print} ($queue.front$)`

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+                \State{Increment frequency count of $ch$ by 1}
+                
+                \While{$Q$ is not empty}
+                    \If{Frequency of top element of $Q$ is greater than 1}
+                        \State{Pop the top element of $Q$}
+                    \Else 
+                        \State{Break this loop}
+                    \EndIf
+                \EndWhile
+                \If{$Q$ is not empty}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: Use `\textbf{not empty, }`

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+                    \State{Print the top element of $Q$}
+                \Else 
+                    \State{Print $null$}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: The format should be **Print**(null)

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            \State{Initialize the $count$ array with 0 for each alphabet}
+            \newline
+            \State{Scan $str$ from left to right and,}
+            \For{each character $ch$}
+                \State{Push $ch$ to $Q$}
+                \State{Increment frequency count of $ch$ by 1}
+                
+                \While{$Q$ is not empty}
+                    \If{Frequency of top element of $Q$ is greater than 1}
+                        \State{Pop the top element of $Q$}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: Use proper notation. `$queue.pop$`

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+                    \Else 
+                        \State{Break this loop}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: Make `Break` boldface.

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex (hunk 2 above)
Comment (helix026) [reply Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: I'm not very sure on how to make changes to an existing pull request, could you please help me out?

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+\captionsetup[algorithm]{labelformat = algnonumber}
+
+\begin{document}
+
+    \begin{algorithm}
+    
+      \caption{Find the first non-repeating character when each character of the stream arrives}
+      
+      \begin{algorithmic}[1]
+        \Ensure {All characters of string $str$ are lowercase alphabets}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: It's better to handle a general case. (Instead of just small case characters).

My suggestion is to use a **HashMap** called **frequency** to store the frequency instead of the count array.

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        \Statex
+        \Function{First\_NonRepeating\_Char}{$str$}
+            \State{Let $Q$ be a queue of characters}
+            \State{Let $count$ be an array of size 26, which keeps track of frequency of alphabets.}
+            \State{Initialize the $count$ array with 0 for each alphabet}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: Create an empty hashmap.

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+            \newline
+            \State{Scan $str$ from left to right and,}
+            \For{each character $ch$}
+                \State{Push $ch$ to $Q$}
+                \State{Increment frequency count of $ch$ by 1}
+                
+                \While{$Q$ is not empty}
+                    \If{Frequency of top element of $Q$ is greater than 1}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: Make it precise.

**Comment by ghost** [comment]
@helix026  I've written some reviews. Please check it out.

You can go through [this](https://github.com/Just-A-Visitor/Algorithmic-Pseudocode/tree/master/Pseudocode/Trees/Vertical%20Order%20Traversal) and [this](https://github.com/Just-A-Visitor/Algorithmic-Pseudocode/tree/master/Pseudocode/Dynamic%20Programming/Wildcard%20Pattern%20Matching) code for more insights.

You can click on the `Files Changed` tab on the right to get an inline view on the comments.

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+                    \EndIf
+                \EndWhile
+                \If{$Q$ is not empty}
+                    \State{Print the top element of $Q$}
+                \Else 
+                    \State{Print $null$}
+                \EndIf
+                
+            \EndFor
+
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: Use `\Statex` to give an empty line before `Return`

**Comment by ghost** [comment]
Remember, the goal is to incorporate the entire algorithm in pseudocode and not just write it in plain English.

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex (hunk 10 above)
Comment (stellarbeam) [reply Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: Should I specify the type of map as (char, int) ?

Comment (ghost) [reply Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: No need. If you want, you can write `\LeftComment{$frequency$ is a hashmap which stores the frequency of the characters}`. (Include this just after the function signature.

**Comment by stellarbeam** [comment]
I've made changes which you suggested. Please review.

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+% Define Left Justified Comments
+\algnewcommand{\LeftComment}[1]{\Statex \(\triangleright\) #1}
+
+% Remove the Numbering of the Algorithm
+\usepackage{caption}
+\DeclareCaptionLabelFormat{algnonumber}{Algorithm}
+\captionsetup[algorithm]{labelformat = algnonumber}
+
+\newcommand{\IsNotEmpty}{\textbf{ is not empty, }}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: You only need a space before the text if you plan to use it inside Math mode. If you are using it in normal mode, then it creates extra space. Hence, the recommended way is to always use it inside Math mode (See the gist).

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+    \begin{algorithm}
+    
+      \caption{Find the first non-repeating character when each character of the stream arrives}
+      
+      \begin{algorithmic}[1]
+        \Statex
+        \Function {$First\_Non\_Repeating\_Char$}{$str$}
+            \LeftComment{$frequency$ is a hashmap which stores the frequency of the characters}
+            \LeftComment{$queue$ is a queue of characters}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: It's obvious, hence don't bother mentioning.

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+            \State{Scan $str$ from left to right and,}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: Not needed if you write the **For Loop** properly.

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+            \For{each character $char$}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: Merge the two lines (See Gist).

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+                \State{$queue$.push($char$)}
+                \State{Increment $frequency$[$char$] by 1}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: Be Precise. A better way is to write it algorithmically, rather than saying it out. Also, I have written a comment (in the gist) to state the assumption that the default frequency of elements which are seen for the first time is zero.

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+                \While{$queue$ \IsNotEmpty}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: As I said, use it inside Math mode to maintain consistency.

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+      \caption{Find the first non-repeating character when each character of the stream arrives}
+      
+      \begin{algorithmic}[1]
+        \Statex
+        \Function {$First\_Non\_Repeating\_Char$}{$str$}
+            \LeftComment{$frequency$ is a hashmap which stores the frequency of the characters}
+            \LeftComment{$queue$ is a queue of characters}
+            \State{Scan $str$ from left to right and,}
+            \For{each character $char$}
+                \State{$queue$.push($char$)}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: I don't see any harm in putting the entire thing `queue.push(char)` inside Math Mode.

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+                \State{Increment $frequency$[$char$] by 1}
+                \While{$queue$ \IsNotEmpty}
+                    \If{$frequency$[$queue.front$] $>$ 1}
+                        \State{$queue.$pop()}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: No need of `()`. It's a pseudocode.

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+                    \Else 
+                        \State{\Break}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: I prefer `\Break Loop`

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+                    \EndIf
+                \EndWhile
+                \If{$queue$ \IsNotEmpty}
+                    \State{\Print($queue.front$)}
+                \Else 
+                    \State{\Print($null)$}
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex]: Explicitly stating that there is no unique character seems better than using `null`.

File: Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+                \EndIf
+                
+            \EndFor
```
Comment (ghost) [inline Pseudocode/Queue/First Non Repeating in Stream/SourceCode.tex:62]: I also added 2 or 3 extra blank lines (using `\Statex` ) to remove the clutter. Since the `for` loop is too long, and hence a bit confusing due to indentation, I removed the `noend` from the package declaration. After that, I manually stopped the `end` for some loops and if-else while retaining it for `For` Loop. Please go through them.

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
@helix026 I have made a lot of changes. The new file can be found [here](https://gist.github.com/Just-A-Visitor/1258dc2d08c1d6e9b3388210aa1f7e00). Make sure that you understand what those changes are (and learn the conventions of this repo in the process). Once you are satisfied with how to do things, Just update your file with my Gist (and resolve the conversation) and ping me. I'll then merge your PR.

**Comment by stellarbeam** [comment]
Thank you so much for being so helpful and patient @Just-A-Visitor , this was my first time and I'm sure I've learnt a lot in the process. :smile:

**Comment by ghost** [comment]
https://github.com/Just-A-Visitor/Algorithmic-Pseudocode/issues/29

Well, Thank you for your contributions. Glad to have you here!
